scripts/pylib/twister/twisterlib.py

```diff
@@ -681,16 +681,19 @@ def monitor_serial(self, ser, halt_fileno, harness):
             # is available yet.
             if serial_line:
                 sl = serial_line.decode('utf-8', 'ignore').lstrip()
```

Comment:
The `.decode()` needs to be inside the `try` block so exception can be caught.

Reply:
Hi @dcpleung, thanks for the review. Actually the exception happens when writing the log, not in decode phasing. If this is the case, do you think I still need to put this line in try block?

Reply:
if that's the case, changing the encoding when the file is opened would probably work.

Reply:
Ok, let me give it a try, thanks!

Another info FYI:  this only happens in ACRN, (actually it happens in grub, I guess grub output some unknow control character, and it cause problem when we are printing it or writing to file).

Reply:
Hi @dcpleung, do you think could I still apply this? I think get a exception and re-open a file with a new encode will need more code to do this. I want just catch exception and skip it because this is not a valid message, and only happened once.

Reply:
TBH, changing the decoding from `utf-8` to `latin-1` here would probably be better. Or open the file as `utf-8` instead. The error comes from mismatched encodings between the opened file, and the line being written to the file. So make them match and the issue should go away. (Note that you will still want to use try/except to catch any error there and print an error message).

Reply:
Thanks so much for your great suggestion, @dcpleung, I tried to implement the you mentioned, it works, like this:
```
diff --git a/scripts/pylib/twister/twisterlib.py b/scripts/pylib/twister/twisterlib.py
index af6ee87f43..474032d77e 100755
--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -645,7 +645,7 @@ class DeviceHandler(Handler):
             harness.handle(None)
             return

-        log_out_fp = open(self.log, "wt")
+        log_out_fp = open(self.log, encoding='utf-8', mode="wt")

         ser_fileno = ser.fileno()
         readlist = [halt_fileno, ser_fileno]
@@ -680,7 +680,11 @@ class DeviceHandler(Handler):
             # Just because ser_fileno has data doesn't mean an entire line
             # is available yet.
             if serial_line:
-                sl = serial_line.decode('utf-8', 'ignore').lstrip()
+                try:
+                    sl = serial_line.decode('utf-8', 'ignore').lstrip()
+                except UnicodeEncodeError:
+                    sl = serial_line.decode('latin-1', 'ignore').lstrip()
+
                 logger.debug("DEVICE: {0}".format(sl.rstrip()))

                 log_out_fp.write(sl)
@@ -2217,8 +2221,8 @@ class CMake():
             results = {"returncode": p.returncode}

         if out:
-            with open(os.path.join(self.build_dir, self.log), "a") as log:
-                log_msg = out.decode(sys.getdefaultencoding())
+            with open(os.path.join(self.build_dir, self.log), encoding='utf-8', mode="a") as log:
+                log_msg = out.decode(sys.getdefaultencoding(), 'ignore')
                 log.write(log_msg)

         return results
```
And during implement this, I was eventually aware of what the problem is. I appreciate the review from you and other reviewers. Sorry the waste your time.